replica-version-map/tla/versionmap.tla

```diff
@@ -0,0 +1,469 @@
+                                               /\ UNCHANGED seqnos_above_local_check_point
+                               ELSE /\ pc' = [pc EXCEPT !["ReplicaEngine"] = "ReplicaLoop"]
+                                    /\ UNCHANGED << lucene, 
+                                                    seqnos_above_local_check_point, 
+                                                    deletion_seqno, 
+                                                    indexing_seqno, 
+                                                    append_only_unsafe_up_to >>
+                         /\ UNCHANGED << isGeneratingRequests, next_seqno, 
+                                         document, request_count, 
+                                         local_check_point >>
+
+AdvanceLocalCheckPoint == /\ pc["ReplicaEngine"] = "AdvanceLocalCheckPoint"
+                          /\ local_check_point' = (CHOOSE n \in (local_check_point .. next_seqno):
+                                                              /\ n > local_check_point
+                                                              /\ (n + 1) \notin seqnos_above_local_check_point
+                                                              /\ ((local_check_point + 2) .. n) \subseteq seqnos_above_local_check_point)
+                          /\ seqnos_above_local_check_point' = { n \in seqnos_above_local_check_point: local_check_point' < n }
+                          /\ IF deletion_seqno /= NULL /\ deletion_seqno <= local_check_point'
+                                THEN /\ deletion_seqno' = NULL
+                                ELSE /\ TRUE
+                                     /\ UNCHANGED deletion_seqno
+                          /\ IF indexing_seqno /= NULL /\ indexing_seqno <= local_check_point'
+                                THEN /\ indexing_seqno' = NULL
+                                ELSE /\ TRUE
+                                     /\ UNCHANGED indexing_seqno
+                          /\ pc' = [pc EXCEPT !["ReplicaEngine"] = "ReplicaLoop"]
+                          /\ UNCHANGED << replication_requests, 
+                                          isGeneratingRequests, lucene, 
+                                          next_seqno, document, request_count, 
+                                          append_only_unsafe_up_to, 
+                                          replication_request >>
+
+ReplicaEngineProcess == ReplicaStart \/ ReplicaLoop
+                           \/ GetReplicationRequest
+                           \/ AdvanceLocalCheckPoint
+
+Next == ReplicationRequestGeneratorProcess \/ LuceneProcess
+           \/ ReplicaEngineProcess
+           \/ (* Disjunct to prevent deadlock on termination *)
+              ((\A self \in ProcSet: pc[self] = "Done") /\ UNCHANGED vars)
+
+Spec == Init /\ [][Next]_vars
+
+Termination == <>(\A self \in ProcSet: pc[self] = "Done")
+
+\* END TRANSLATION
```

Comment:
Do we want some more invariants below? At the moment they're mostly about properties at termination, which is simple and ought to be sufficient, but we could have more if desired.